# Test

## Service

### New Cluster

```
2026/01/27 17:14:41.604779 [DEBUG] common/rpc/client_lb.go:251 [8d3ae9fd7e925889:8a73bc4697aa1c48:VolumeDegrade.Initialize:rpc.lb.doCtx] to start /kv/get/volume_degrade_stats
2026/01/27 17:14:41.606065 [ERROR] scheduler/volume_degrade_stats.go:359 [8d3ae9fd7e925889:8a73bc4697aa1c48] failed to get volume degrade stats from kv, err[Not Found]
2026/01/27 17:14:41.606074 [INFO] scheduler/volume_degrade_stats.go:327 [8d3ae9fd7e925889:8a73bc4697aa1c48] succeed to get all exist stats from kv
2026/01/27 17:14:41.606093 [INFO] scheduler/volume_degrade_stats.go:340 [8d3ae9fd7e925889:8a73bc4697aa1c48] succeed to initialized: batchSize[50], stats[&{batchSize:50 blobCounter:map[] volCounter:map[]}]
```

### Restart with New Batch

```
2026/01/27 17:17:34.411720 [INFO] scheduler/volume_degrade_stats.go:362 [079e1e151a8195cf:a691f05ebf5386bb] succeed to get volume degrade stats[&{BatchSize:50 BlobStats:[{BatchIndex:0 CodeModeStats:[{CodeModeName:EC3P3 DegradeStats:[{Level:1 Count:32}]}]} {BatchIndex:1 CodeModeStats:[{CodeModeName:EC3P3 DegradeStats:[{Level:1 Count:240}]}]}] VolStats:[{BatchIndex:0 CodeModeStats:[{CodeModeName:EC3P3 DegradeStats:[{Level:1 Count:2}]}]} {BatchIndex:1 CodeModeStats:[{CodeModeName:EC3P3 DegradeStats:[{Level:1 Count:8}]}]}]}] from kv
2026/01/27 17:17:34.411746 [INFO] scheduler/volume_degrade_stats.go:327 [079e1e151a8195cf:a691f05ebf5386bb] succeed to get all exist stats from kv
2026/01/27 17:17:34.411757 [INFO] scheduler/volume_degrade_stats.go:331 [079e1e151a8195cf:a691f05ebf5386bb] degrade batch changed from 50 to 10, will delete all exist stats
2026/01/27 17:17:34.411781 [DEBUG] common/rpc/client_lb.go:251 [079e1e151a8195cf:a691f05ebf5386bb:VolumeDegrade.Initialize:rpc.lb.doCtx] to start /kv/delete/volume_degrade_stats
2026/01/27 17:17:34.413314 [INFO] scheduler/volume_degrade_stats.go:373 [079e1e151a8195cf:a691f05ebf5386bb] succeed to delete volume degrade stats from kv
2026/01/27 17:17:34.413325 [INFO] scheduler/volume_degrade_stats.go:336 [079e1e151a8195cf:a691f05ebf5386bb] succeed to delete all exist stats from clustermgr
2026/01/27 17:17:34.413352 [INFO] scheduler/volume_degrade_stats.go:340 [079e1e151a8195cf:a691f05ebf5386bb] succeed to initialized: batchSize[10], stats[&{batchSize:10 blobCounter:map[] volCounter:map[]}]
```

### Restart

```
2026/01/27 17:21:39.594193 [DEBUG] common/rpc/client_lb.go:251 [9196fc0bab65957c:312b56fb0f4883b7:VolumeDegrade.Initialize:rpc.lb.doCtx] to start /kv/get/volume_degrade_stats
2026/01/27 17:21:39.596017 [INFO] scheduler/volume_degrade_stats.go:362 [9196fc0bab65957c:312b56fb0f4883b7] succeed to get volume degrade stats[&{BatchSize:10 BlobStats:[{BatchIndex:4 CodeModeStats:[{CodeModeName:EC3P3 DegradeStats:[{Level:1 Count:32}]}]} {BatchIndex:5 CodeModeStats:[{CodeModeName:EC3P3 DegradeStats:[{Level:1 Count:218}]}]} {BatchIndex:6 CodeModeStats:[{CodeModeName:EC3P3 DegradeStats:[{Level:1 Count:22}]}]}] VolStats:[{BatchIndex:4 CodeModeStats:[{CodeModeName:EC3P3 DegradeStats:[{Level:1 Count:2}]}]} {BatchIndex:5 CodeModeStats:[{CodeModeName:EC3P3 DegradeStats:[{Level:1 Count:7}]}]} {BatchIndex:6 CodeModeStats:[{CodeModeName:EC3P3 DegradeStats:[{Level:1 Count:1}]}]}]}] from kv
2026/01/27 17:21:39.596133 [INFO] scheduler/volume_degrade_stats.go:327 [9196fc0bab65957c:312b56fb0f4883b7] succeed to get all exist stats from kv
2026/01/27 17:21:39.596189 [INFO] scheduler/volume_degrade_stats.go:340 [9196fc0bab65957c:312b56fb0f4883b7] succeed to initialized: batchSize[10], stats[&{batchSize:10 blobCounter:map[4:0xc000124e30 5:0xc000124e40 6:0xc000124e50] volCounter:map[4:0xc000124e60 5:0xc000124e70 6:0xc000124e80]}]
```

### Metrics

```
[root@szjd-yfq-pm-os01-bconest-4 cubefs (yzq-wip-volume-degrade-stats)] # curl -qs http://127.0.0.1:9800/metrics | grep degrade
# HELP blobstore_scheduler_degrade_stats scheduler degrade stats
# TYPE blobstore_scheduler_degrade_stats gauge
blobstore_scheduler_degrade_stats{cluster_id="0",codename="EC3P3",degrade_level="1",item="blob"} 32
blobstore_scheduler_degrade_stats{cluster_id="0",codename="EC3P3",degrade_level="1",item="volume"} 2
[root@szjd-yfq-pm-os01-bconest-4 cubefs (yzq-wip-volume-degrade-stats)] #
```

## Cli

### help

```
[root@szjd-yfq-pm-os01-bconest-4 cubefs (yzq-wip-volume-degrade-stats)] # ./build/bin/blobstore/blobstore-cli scheduler -h

scheduler tools

Usage:
  scheduler [flags]

Flags:
  -h, --help     display help

Sub Commands:
  checkpoint  inspect checkpoint tools
  degrade     degrade stats tools
  kafka       kafka consume tools
  migrate     migrate tools
  stat        show leader stat of scheduler

[root@szjd-yfq-pm-os01-bconest-4 cubefs (yzq-wip-volume-degrade-stats)] # ./build/bin/blobstore/blobstore-cli scheduler degrade -h

degrade stats tools for scheduler

Usage:
  degrade [flags]

Flags:
  -h, --help     display help

Sub Commands:
  all     get blob and volume degrade stats
  blob    get blob degrade stats
  volume  get volume degrade stats

[root@szjd-yfq-pm-os01-bconest-4 cubefs (yzq-wip-volume-degrade-stats)] #
```

### Result

```
[root@szjd-yfq-pm-os01-bconest-4 cubefs (yzq-wip-volume-degrade-stats)] # ./build/bin/blobstore/blobstore-cli scheduler degrade all
2026/01/27 19:14:12.715802 [DEBUG] scheduler/client/clustermgr.go:1285 [d9153eabadc99277-cmd:6b7e80c36a106fa1] getting volume degrade stats
2026/01/27 19:14:12.715971 [DEBUG] common/rpc/client_lb.go:251 [d9153eabadc99277-cmd:6b7e80c36a106fa1:from-cmd:rpc.lb.doCtx] to start /kv/get/volume_degrade_stats
2026/01/27 19:14:12.717777 [DEBUG] common/rpc/client_lb.go:321 [d9153eabadc99277-cmd:6b7e80c36a106fa1:from-cmd:rpc.lb.doCtx] the last host of request, try-times:1 code:200 host:(1 http://127.0.0.1:9998) <nil>
{
    "blob": [
        {
            "name": "EC3P3",
            "degrade": [
                {
                    "level": 1,
                    "count": 272
                }
            ]
        }
    ],
    "volume": [
        {
            "name": "EC3P3",
            "degrade": [
                {
                    "level": 1,
                    "count": 10
                }
            ]
        }
    ]
}
[root@szjd-yfq-pm-os01-bconest-4 cubefs (yzq-wip-volume-degrade-stats)] # ./build/bin/blobstore/blobstore-cli scheduler degrade blob
2026/01/27 19:14:16.052296 [DEBUG] scheduler/client/clustermgr.go:1285 [efb1d39e421e6459-cmd:7fdeb2405700db66] getting volume degrade stats
2026/01/27 19:14:16.052515 [DEBUG] common/rpc/client_lb.go:251 [efb1d39e421e6459-cmd:7fdeb2405700db66:from-cmd:rpc.lb.doCtx] to start /kv/get/volume_degrade_stats
2026/01/27 19:14:16.054695 [DEBUG] common/rpc/client_lb.go:321 [efb1d39e421e6459-cmd:7fdeb2405700db66:from-cmd:rpc.lb.doCtx] the last host of request, try-times:1 code:200 host:(1 http://127.0.0.1:9998) <nil>
{
    "blob": [
        {
            "name": "EC3P3",
            "degrade": [
                {
                    "level": 1,
                    "count": 272
                }
            ]
        }
    ]
}
[root@szjd-yfq-pm-os01-bconest-4 cubefs (yzq-wip-volume-degrade-stats)] # ./build/bin/blobstore/blobstore-cli scheduler degrade volume
2026/01/27 19:14:19.431837 [DEBUG] scheduler/client/clustermgr.go:1285 [8d15c9ae8d11a8f5-cmd:2101846bab888e95] getting volume degrade stats
2026/01/27 19:14:19.431966 [DEBUG] common/rpc/client_lb.go:251 [8d15c9ae8d11a8f5-cmd:2101846bab888e95:from-cmd:rpc.lb.doCtx] to start /kv/get/volume_degrade_stats
2026/01/27 19:14:19.433802 [DEBUG] common/rpc/client_lb.go:321 [8d15c9ae8d11a8f5-cmd:2101846bab888e95:from-cmd:rpc.lb.doCtx] the last host of request, try-times:1 code:200 host:(1 http://127.0.0.1:9998) <nil>
{
    "volume": [
        {
            "name": "EC3P3",
            "degrade": [
                {
                    "level": 1,
                    "count": 10
                }
            ]
        }
    ]
}
[root@szjd-yfq-pm-os01-bconest-4 cubefs (yzq-wip-volume-degrade-stats)] #
```

--- a/blobstore/cli/scheduler/scheduler.go
+++ b/blobstore/cli/scheduler/scheduler.go
@@ -84,6 +84,7 @@ func Register(app *grumble.App) {
 	addCmdMigrateTask(schedulerCommand)
 	addCmdVolumeInspectCheckpointTask(schedulerCommand)
 	addCmdKafkaConsumer(schedulerCommand)
+	addCmdVolumeDegradeStats(schedulerCommand)
 }
 
 func leaderStat(c *grumble.Context) error {

--- a/blobstore/cli/scheduler/volume_inspect.go
+++ b/blobstore/cli/scheduler/volume_inspect.go
@@ -70,3 +70,125 @@ func cmdSetInspectCheckpoint(c *grumble.Context) error {
 	fmt.Println("set volume inspect checkpoint successfully")
 	return nil
 }
+
+func addCmdVolumeDegradeStats(cmd *grumble.Command) {
+	degradeStatsCommand := &grumble.Command{
+		Name:     "degrade",
+		Help:     "degrade stats tools",
+		LongHelp: "degrade stats tools for scheduler",
+	}
+	cmd.AddCommand(degradeStatsCommand)
+
+	degradeStatsCommand.AddCommand(&grumble.Command{
+		Name:  "all",
+		Help:  "get blob and volume degrade stats",
+		Run:   cmdGetDegradeStatsAll,
+		Flags: clusterFlags,
+	})
+	degradeStatsCommand.AddCommand(&grumble.Command{
+		Name:  "blob",
+		Help:  "get blob degrade stats",
+		Run:   cmdGetDegradeStatsBlob,
+		Flags: clusterFlags,
+	})
+	degradeStatsCommand.AddCommand(&grumble.Command{
+		Name:  "volume",
+		Help:  "get volume degrade stats",
+		Run:   cmdGetDegradeStatsVolume,
+		Flags: clusterFlags,
+	})
+}
+
+type aggregatedLevel struct {
+	Level int    `json:"level"`
+	Count uint32 `json:"count"`
+}
+
+type aggregatedCodeMode struct {
+	CodeMode     string            `json:"codemode"`
+	DegradeStats []aggregatedLevel `json:"degrade"`
+}
+
+type aggregatedDegradeStat struct {
+	BlobStats   []aggregatedCodeMode `json:"blob"`
+	VolumeStats []aggregatedCodeMode `json:"volume"`
+}
+
+func getDegradeStatsFromKV(c *grumble.Context) (*aggregatedDegradeStat, error) {
+	clusterMgrCli := newClusterMgrTaskClient(getClusterID(c.Flags))
+	stats, err := clusterMgrCli.GetVolumeDegradeStats(common.CmdContext())
+	if err != nil {
+		return nil, err
+	}
+
+	aggregate := func(batches []proto.VolumeDegradeBatch) []aggregatedCodeMode {
+		agg := make(map[string]map[int]uint32)
+
+		for _, batch := range batches {
+			for _, cm := range batch.CodeModeStats {
+				levelMap, ok := agg[cm.Mode.String()]
+				if !ok {
+					levelMap = make(map[int]uint32)
+					agg[cm.Mode.String()] = levelMap
+				}
+				for _, level := range cm.DegradeStats {
+					levelMap[level.Level] += level.Count
+				}
+			}
+		}
+
+		result := make([]aggregatedCodeMode, 0, len(agg))
+		for name, levelMap := range agg {
+			degradeStats := make([]aggregatedLevel, 0, len(levelMap))
+			for level, count := range levelMap {
+				degradeStats = append(degradeStats, aggregatedLevel{
+					Level: level,
+					Count: count,
+				})
+			}
+			result = append(result, aggregatedCodeMode{
+				CodeMode:     name,
+				DegradeStats: degradeStats,
+			})
+		}
+		return result
+	}
+
+	return &aggregatedDegradeStat{
+		BlobStats:   aggregate(stats.BlobStats),
+		VolumeStats: aggregate(stats.VolStats),
+	}, nil
+}
+
+func cmdGetDegradeStatsAll(c *grumble.Context) error {
+	stats, err := getDegradeStatsFromKV(c)
+	if err != nil {
+		return err
+	}
+	fmt.Println(common.Readable(stats))
+	return nil
+}
+
+func cmdGetDegradeStatsBlob(c *grumble.Context) error {
+	stats, err := getDegradeStatsFromKV(c)
+	if err != nil {
+		return err
+	}
+	output := struct {
+		BlobStats []aggregatedCodeMode `json:"blob"`
+	}{BlobStats: stats.BlobStats}
+	fmt.Println(common.Readable(output))
+	return nil
+}
+
+func cmdGetDegradeStatsVolume(c *grumble.Context) error {
+	stats, err := getDegradeStatsFromKV(c)
+	if err != nil {
+		return err
+	}
+	output := struct {
+		VolumeStats []aggregatedCodeMode `json:"volume"`
+	}{VolumeStats: stats.VolumeStats}
+	fmt.Println(common.Readable(output))
+	return nil
+}

--- a/blobstore/common/proto/scheduler.go
+++ b/blobstore/common/proto/scheduler.go
@@ -313,6 +313,27 @@ type VolumeInspectRet struct {
 	MissedShards  []*MissedShard `json:"missed_shards"`
 }
 
+type VolumeDegradeLevel struct {
+	Level int    `json:"level"`
+	Count uint32 `json:"count"`
+}
+
+type VolumeDegradeCodeMode struct {
+	Mode         codemode.CodeMode    `json:"mode"`
+	DegradeStats []VolumeDegradeLevel `json:"degrade"`
+}
+
+type VolumeDegradeBatch struct {
+	BatchIndex    uint32                  `json:"batch_index"`
+	CodeModeStats []VolumeDegradeCodeMode `json:"codemode"`
+}
+
+type VolumeDegradeStats struct {
+	BatchSize uint32               `json:"batch_size"`
+	BlobStats []VolumeDegradeBatch `json:"blob"`
+	VolStats  []VolumeDegradeBatch `json:"volume"`
+}
+
 func (inspect *VolumeInspectRet) Err() error {
 	if len(inspect.InspectErrStr) == 0 {
 		return nil

--- a/blobstore/scheduler/client/clustermgr.go
+++ b/blobstore/scheduler/client/clustermgr.go
@@ -98,6 +98,9 @@ type ClusterMgrTaskAPI interface {
 	SetVolumeInspectCheckPoint(ctx context.Context, startVid proto.Vid) (err error)
 	GetConsumeOffset(taskType proto.TaskType, topic string, partition int32) (offset int64, err error)
 	SetConsumeOffset(taskType proto.TaskType, topic string, partition int32, offset int64) (err error)
+	GetVolumeDegradeStats(ctx context.Context) (stats *proto.VolumeDegradeStats, err error)
+	SetVolumeDegradeStats(ctx context.Context, stats *proto.VolumeDegradeStats) (err error)
+	DeleteVolumeDegradeStats(ctx context.Context) (err error)
 }
 
 // ClusterMgrAPI define the interface of clustermgr used by scheduler
@@ -149,6 +152,7 @@ const (
 	_migratingDiskPrefix = "migrating"
 	_checkPoint          = "checkpoint"
 	_consumeOffset       = "consume_offset"
+	_volumeDegradeStats  = "volume_degrade_stats"
 )
 
 var (
@@ -1272,6 +1276,52 @@ func (c *clustermgrClient) listAllShardDisks(ctx context.Context, status proto.D
 			break
 		}
 		args.Marker = selectMarker
+	}
+	return
+}
+
+func (c *clustermgrClient) GetVolumeDegradeStats(ctx context.Context) (*proto.VolumeDegradeStats, error) {
+	span := trace.SpanFromContextSafe(ctx)
+	span.Debugf("getting volume degrade stats")
+
+	ret, err := c.client.GetKV(ctx, _volumeDegradeStats)
+	if err != nil {
+		span.Errorf("failed to get volume degrade stats from kv: err[%+v]", err)
+		return nil, err
+	}
+
+	var stats proto.VolumeDegradeStats
+	if err := json.Unmarshal(ret.Value, &stats); err != nil {
+		span.Errorf("failed to unmarshal stats: err[%+v]", err)
+		return nil, err
+	}
+
+	return &stats, nil
+}
+
+func (c *clustermgrClient) SetVolumeDegradeStats(ctx context.Context, stats *proto.VolumeDegradeStats) (err error) {
+	span := trace.SpanFromContextSafe(ctx)
+	span.Debugf("setting volume degrade stats: [%+v]", stats)
+
+	statsBytes, err := json.Marshal(stats)
+	if err != nil {
+		span.Errorf("failed to marshal volume degrade stats: err[%+v]", err)
+		return
+	}
+	err = c.client.SetKV(ctx, _volumeDegradeStats, statsBytes)
+	if err != nil {
+		span.Errorf("failed to set volume degrade stats to kv: err[%+v]", err)
+	}
+	return
+}
+
+func (c *clustermgrClient) DeleteVolumeDegradeStats(ctx context.Context) (err error) {
+	span := trace.SpanFromContextSafe(ctx)
+	span.Debugf("deleting volume degrade stats")
+
+	err = c.client.DeleteKV(ctx, _volumeDegradeStats)
+	if err != nil {
+		span.Errorf("failed to delete volume degrade stats from kv: err[%+v]", err)
 	}
 	return
 }

--- a/blobstore/scheduler/client/clustermgr_test.go
+++ b/blobstore/scheduler/client/clustermgr_test.go
@@ -480,4 +480,75 @@ func TestClustermgrClient(t *testing.T) {
 		require.NoError(t, err)
 		require.Equal(t, offset, offset2)
 	}
+	{
+		stats := &proto.VolumeDegradeStats{
+			BatchSize: 10000,
+			BlobStats: []proto.VolumeDegradeBatch{
+				{
+					BatchIndex: 1,
+					CodeModeStats: []proto.VolumeDegradeCodeMode{
+						{
+							Mode: codemode.EC3P3,
+							DegradeStats: []proto.VolumeDegradeLevel{
+								{
+									Level: 1,
+									Count: 123,
+								},
+							},
+						},
+					},
+				},
+			},
+			VolStats: []proto.VolumeDegradeBatch{
+				{
+					BatchIndex: 1,
+					CodeModeStats: []proto.VolumeDegradeCodeMode{
+						{
+							Mode: codemode.EC3P3,
+							DegradeStats: []proto.VolumeDegradeLevel{
+								{
+									Level: 1,
+									Count: 123,
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		// set degrade stats failed
+		cli.client.(*MockClusterManager).EXPECT().SetKV(any, any, any).Return(errMock)
+		err := cli.SetVolumeDegradeStats(ctx, stats)
+		require.NotNil(t, err)
+
+		// set degrade stats success
+		cli.client.(*MockClusterManager).EXPECT().SetKV(any, any, any).Return(nil)
+		err = cli.SetVolumeDegradeStats(ctx, stats)
+		require.NoError(t, err)
+
+		// get degrade stats failed
+		cli.client.(*MockClusterManager).EXPECT().GetKV(any, any).Return(cmapi.GetKvRet{}, errMock)
+		_, err = cli.GetVolumeDegradeStats(ctx)
+		require.NotNil(t, err)
+
+		// get degrade stats success
+		statsBytes, _ := json.Marshal(stats)
+		cli.client.(*MockClusterManager).EXPECT().GetKV(any, any).Return(cmapi.GetKvRet{Value: statsBytes}, nil)
+		statsRet, err := cli.GetVolumeDegradeStats(ctx)
+		require.NoError(t, err)
+		require.Equal(t, statsRet.BatchSize, stats.BatchSize)
+		require.Equal(t, statsRet.BlobStats[0].BatchIndex, stats.BlobStats[0].BatchIndex)
+		require.Equal(t, statsRet.VolStats[0].BatchIndex, stats.VolStats[0].BatchIndex)
+
+		// delete degrade stats failed
+		cli.client.(*MockClusterManager).EXPECT().DeleteKV(any, any).Return(errMock)
+		err = cli.DeleteVolumeDegradeStats(ctx)
+		require.NotNil(t, err)
+
+		// delete degrade stats success
+		cli.client.(*MockClusterManager).EXPECT().DeleteKV(any, any).Return(nil)
+		err = cli.DeleteVolumeDegradeStats(ctx)
+		require.NoError(t, err)
+	}
 }

--- a/blobstore/scheduler/client_mock_test.go
+++ b/blobstore/scheduler/client_mock_test.go
@@ -123,6 +123,20 @@ func (mr *MockClusterMgrAPIMockRecorder) DeleteMigratingDisk(arg0, arg1, arg2 in
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "DeleteMigratingDisk", reflect.TypeOf((*MockClusterMgrAPI)(nil).DeleteMigratingDisk), arg0, arg1, arg2)
 }
 
+// DeleteVolumeDegradeStats mocks base method.
+func (m *MockClusterMgrAPI) DeleteVolumeDegradeStats(arg0 context.Context) error {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "DeleteVolumeDegradeStats", arg0)
+	ret0, _ := ret[0].(error)
+	return ret0
+}
+
+// DeleteVolumeDegradeStats indicates an expected call of DeleteVolumeDegradeStats.
+func (mr *MockClusterMgrAPIMockRecorder) DeleteVolumeDegradeStats(arg0 interface{}) *gomock.Call {
+	mr.mock.ctrl.T.Helper()
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "DeleteVolumeDegradeStats", reflect.TypeOf((*MockClusterMgrAPI)(nil).DeleteVolumeDegradeStats), arg0)
+}
+
 // GetConfig mocks base method.
 func (m *MockClusterMgrAPI) GetConfig(arg0 context.Context, arg1 string) (string, error) {
 	m.ctrl.T.Helper()
@@ -241,6 +255,21 @@ func (m *MockClusterMgrAPI) GetShardInfo(arg0 context.Context, arg1 proto.ShardI
 func (mr *MockClusterMgrAPIMockRecorder) GetShardInfo(arg0, arg1 interface{}) *gomock.Call {
 	mr.mock.ctrl.T.Helper()
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetShardInfo", reflect.TypeOf((*MockClusterMgrAPI)(nil).GetShardInfo), arg0, arg1)
+}
+
+// GetVolumeDegradeStats mocks base method.
+func (m *MockClusterMgrAPI) GetVolumeDegradeStats(arg0 context.Context) (*proto.VolumeDegradeStats, error) {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "GetVolumeDegradeStats", arg0)
+	ret0, _ := ret[0].(*proto.VolumeDegradeStats)
+	ret1, _ := ret[1].(error)
+	return ret0, ret1
+}
+
+// GetVolumeDegradeStats indicates an expected call of GetVolumeDegradeStats.
+func (mr *MockClusterMgrAPIMockRecorder) GetVolumeDegradeStats(arg0 interface{}) *gomock.Call {
+	mr.mock.ctrl.T.Helper()
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetVolumeDegradeStats", reflect.TypeOf((*MockClusterMgrAPI)(nil).GetVolumeDegradeStats), arg0)
 }
 
 // GetVolumeInfo mocks base method.
@@ -639,6 +668,20 @@ func (m *MockClusterMgrAPI) SetShardDiskRepairing(arg0 context.Context, arg1 pro
 func (mr *MockClusterMgrAPIMockRecorder) SetShardDiskRepairing(arg0, arg1 interface{}) *gomock.Call {
 	mr.mock.ctrl.T.Helper()
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "SetShardDiskRepairing", reflect.TypeOf((*MockClusterMgrAPI)(nil).SetShardDiskRepairing), arg0, arg1)
+}
+
+// SetVolumeDegradeStats mocks base method.
+func (m *MockClusterMgrAPI) SetVolumeDegradeStats(arg0 context.Context, arg1 *proto.VolumeDegradeStats) error {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "SetVolumeDegradeStats", arg0, arg1)
+	ret0, _ := ret[0].(error)
+	return ret0
+}
+
+// SetVolumeDegradeStats indicates an expected call of SetVolumeDegradeStats.
+func (mr *MockClusterMgrAPIMockRecorder) SetVolumeDegradeStats(arg0, arg1 interface{}) *gomock.Call {
+	mr.mock.ctrl.T.Helper()
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "SetVolumeDegradeStats", reflect.TypeOf((*MockClusterMgrAPI)(nil).SetVolumeDegradeStats), arg0, arg1)
 }
 
 // SetVolumeInspectCheckPoint mocks base method.

--- a/blobstore/scheduler/metric.go
+++ b/blobstore/scheduler/metric.go
@@ -1,0 +1,44 @@
+package scheduler
+
+import (
+	"strconv"
+
+	"github.com/cubefs/cubefs/blobstore/common/codemode"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	_metricDegradeStatsItemBlob   = "blob"
+	_metricDegradeStatsItemVolume = "volume"
+)
+
+var dgradeStatsMetric = prometheus.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Namespace: "blobstore",
+		Subsystem: "scheduler",
+		Name:      "degrade_stats",
+		Help:      "scheduler degrade stats",
+	},
+	[]string{"cluster_id", "item", "codename", "degrade_level"},
+)
+
+func init() {
+	prometheus.MustRegister(dgradeStatsMetric)
+}
+
+func reportDegradeStats(
+	clusterID string,
+	item string,
+	aggregated map[codemode.CodeMode]map[int]float64,
+) {
+	for cm, levelMap := range aggregated {
+		for level, total := range levelMap {
+			dgradeStatsMetric.WithLabelValues(
+				clusterID,
+				item,
+				cm.String(),
+				strconv.Itoa(level),
+			).Set(total)
+		}
+	}
+}

--- a/blobstore/scheduler/startup.go
+++ b/blobstore/scheduler/startup.go
@@ -194,6 +194,9 @@ func NewService(conf *Config) (svr *Service, err error) {
 		return nil, err
 	}
 	inspectMgr := NewVolumeInspectMgr(clusterMgrCli, mqProxy, inspectorTaskSwitch, &conf.VolumeInspect)
+	if err = inspectMgr.degradeStats.initialize(); err != nil {
+		return nil, err
+	}
 
 	//===========shard module migrate manager===============
 	// new shard disk repair manager

--- a/blobstore/scheduler/volume_degrade_stats.go
+++ b/blobstore/scheduler/volume_degrade_stats.go
@@ -1,0 +1,375 @@
+// Copyright 2022 The CubeFS Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package scheduler
+
+import (
+	"context"
+	"sort"
+	"strings"
+
+	"github.com/cubefs/cubefs/blobstore/common/codemode"
+	errcode "github.com/cubefs/cubefs/blobstore/common/errors"
+	"github.com/cubefs/cubefs/blobstore/common/proto"
+	"github.com/cubefs/cubefs/blobstore/common/trace"
+	"github.com/cubefs/cubefs/blobstore/scheduler/client"
+)
+
+type degradeCodeModeCounter struct {
+	codeMode       codemode.CodeMode
+	degradeCounter map[int]uint32
+}
+
+type degradeBatchCounter struct {
+	batchIndex      uint32
+	codeModeCounter map[codemode.CodeMode]*degradeCodeModeCounter
+}
+
+func newBatchCounter(batchIndex uint32) *degradeBatchCounter {
+	return &degradeBatchCounter{
+		batchIndex:      batchIndex,
+		codeModeCounter: make(map[codemode.CodeMode]*degradeCodeModeCounter),
+	}
+}
+
+type degradeStatsCounter struct {
+	batchSize   uint32
+	blobCounter map[uint32]*degradeBatchCounter
+	volCounter  map[uint32]*degradeBatchCounter
+}
+
+func newDegradeStatsCounter(batchSize uint32) *degradeStatsCounter {
+	return &degradeStatsCounter{
+		batchSize:   batchSize,
+		blobCounter: make(map[uint32]*degradeBatchCounter),
+		volCounter:  make(map[uint32]*degradeBatchCounter),
+	}
+}
+
+func degradeStatsToRecord(stats map[uint32]*degradeBatchCounter) []proto.VolumeDegradeBatch {
+	record := make([]proto.VolumeDegradeBatch, 0, len(stats))
+
+	batchIndices := make([]uint32, 0, len(stats))
+	for idx := range stats {
+		batchIndices = append(batchIndices, idx)
+	}
+	sort.Slice(batchIndices, func(i, j int) bool {
+		return batchIndices[i] < batchIndices[j]
+	})
+
+	for _, batchIndex := range batchIndices {
+		batchCounter := stats[batchIndex]
+		batchProto := proto.VolumeDegradeBatch{
+			BatchIndex:    batchIndex,
+			CodeModeStats: make([]proto.VolumeDegradeCodeMode, 0, len(batchCounter.codeModeCounter)),
+		}
+
+		codeModeNames := make([]string, 0, len(batchCounter.codeModeCounter))
+		codeModeMap := make(map[string]codemode.CodeMode)
+		for cm := range batchCounter.codeModeCounter {
+			name := cm.String()
+			codeModeNames = append(codeModeNames, name)
+			codeModeMap[name] = cm
+		}
+		sort.Strings(codeModeNames)
+
+		for _, name := range codeModeNames {
+			cm := codeModeMap[name]
+			codeModeCounter := batchCounter.codeModeCounter[cm]
+
+			degradeStats := make([]proto.VolumeDegradeLevel, 0, len(codeModeCounter.degradeCounter))
+			levels := make([]int, 0, len(codeModeCounter.degradeCounter))
+			for level := range codeModeCounter.degradeCounter {
+				levels = append(levels, level)
+			}
+			sort.Ints(levels)
+
+			for _, level := range levels {
+				degradeStats = append(degradeStats, proto.VolumeDegradeLevel{
+					Level: level,
+					Count: codeModeCounter.degradeCounter[level],
+				})
+			}
+
+			batchProto.CodeModeStats = append(batchProto.CodeModeStats, proto.VolumeDegradeCodeMode{
+				Mode:         cm,
+				DegradeStats: degradeStats,
+			})
+		}
+
+		record = append(record, batchProto)
+	}
+
+	return record
+}
+
+func degradeStatsFromRecord(record []proto.VolumeDegradeBatch) map[uint32]*degradeBatchCounter {
+	counter := make(map[uint32]*degradeBatchCounter)
+
+	for _, batchProto := range record {
+		batchIndex := batchProto.BatchIndex
+		batchCounter, exists := counter[batchIndex]
+		if !exists {
+			batchCounter = &degradeBatchCounter{
+				batchIndex:      batchIndex,
+				codeModeCounter: make(map[codemode.CodeMode]*degradeCodeModeCounter),
+			}
+			counter[batchIndex] = batchCounter
+		}
+
+		for _, codeModeProto := range batchProto.CodeModeStats {
+			codeMode := codeModeProto.Mode
+
+			codeModeCounter, exists := batchCounter.codeModeCounter[codeMode]
+			if !exists {
+				codeModeCounter = &degradeCodeModeCounter{
+					codeMode:       codeMode,
+					degradeCounter: make(map[int]uint32),
+				}
+				batchCounter.codeModeCounter[codeMode] = codeModeCounter
+			}
+
+			for _, levelStat := range codeModeProto.DegradeStats {
+				codeModeCounter.degradeCounter[levelStat.Level] += levelStat.Count
+			}
+		}
+	}
+
+	return counter
+}
+
+func (b *degradeStatsCounter) fromRecord(record *proto.VolumeDegradeStats) {
+	b.batchSize = record.BatchSize
+	b.blobCounter = degradeStatsFromRecord(record.BlobStats)
+	b.volCounter = degradeStatsFromRecord(record.VolStats)
+}
+
+func (b *degradeStatsCounter) toRecord() *proto.VolumeDegradeStats {
+	record := &proto.VolumeDegradeStats{
+		BatchSize: b.batchSize,
+	}
+	record.BlobStats = degradeStatsToRecord(b.blobCounter)
+	record.VolStats = degradeStatsToRecord(b.volCounter)
+	return record
+}
+
+type VolumeDegradeStatsCfg struct {
+	Enable bool   `json:"enable"`
+	Batch  uint32 `json:"batch"`
+}
+
+type VolumeDegradeStatsMgr struct {
+	clusterID proto.ClusterID
+
+	stats         *degradeStatsCounter
+	curBlobStats  *degradeBatchCounter
+	curVolStats   *degradeBatchCounter
+	curBatchIndex uint32
+	firstUpdate   bool
+
+	clusterMgrCli client.ClusterMgrAPI
+	cfg           *VolumeDegradeStatsCfg
+}
+
+func newVolumeDegradeStatsMgr(
+	clusterID proto.ClusterID,
+	clusterMgrCli client.ClusterMgrAPI,
+	cfg *VolumeDegradeStatsCfg,
+) *VolumeDegradeStatsMgr {
+	return &VolumeDegradeStatsMgr{
+		clusterID:     clusterID,
+		stats:         newDegradeStatsCounter(cfg.Batch),
+		firstUpdate:   true,
+		clusterMgrCli: clusterMgrCli,
+		cfg:           cfg,
+	}
+}
+
+// When an error is encountered, only log it to avoid affecting
+// the original volume inspect logic.
+func (v *VolumeDegradeStatsMgr) updateDegradeInfo(
+	ctx context.Context,
+	vid uint32,
+	codeMode codemode.CodeMode,
+	bidsMissed map[proto.BlobID][]uint8,
+) {
+	span := trace.SpanFromContextSafe(ctx)
+
+	if !v.cfg.Enable {
+		span.Debugf("volume degrade stats don't enabled, just return")
+		return
+	}
+
+	batchIndex := v.getBatchIndex(vid)
+
+	// Lazy initialization curStats
+	if v.firstUpdate {
+		v.firstUpdate = false
+		v.curBatchIndex = batchIndex
+		v.curBlobStats = newBatchCounter(batchIndex)
+		v.curVolStats = newBatchCounter(batchIndex)
+	}
+
+	// Determine whether to switch to a new batch
+	if batchIndex != v.curBatchIndex {
+		span.Infof("batch index changed from %d to %d", v.curBatchIndex, batchIndex)
+
+		// 1. update stats
+		v.stats.blobCounter[v.curBatchIndex] = v.curBlobStats
+		v.stats.volCounter[v.curBatchIndex] = v.curVolStats
+
+		// 2. create a new batch
+		v.curBatchIndex = batchIndex
+		v.curBlobStats = newBatchCounter(batchIndex)
+		v.curVolStats = newBatchCounter(batchIndex)
+
+		// 3. set to kv
+		if err := v.setStatsToKV(ctx); err != nil {
+			return
+		}
+
+		// 4. update prometheus
+		v.updateMetric()
+	}
+
+	// Update curBatchStats
+	maxMissingShard := 0
+	getOrCreateCounter := func(dbc *degradeBatchCounter) *degradeCodeModeCounter {
+		if counter, exists := dbc.codeModeCounter[codeMode]; exists {
+			return counter
+		}
+
+		counter := &degradeCodeModeCounter{
+			codeMode:       codeMode,
+			degradeCounter: make(map[int]uint32),
+		}
+		dbc.codeModeCounter[codeMode] = counter
+		return counter
+	}
+
+	blobCounter := getOrCreateCounter(v.curBlobStats)
+	for _, shardMissing := range bidsMissed {
+		degradeLevel := len(shardMissing)
+		if maxMissingShard < degradeLevel {
+			maxMissingShard = degradeLevel
+		}
+		blobCounter.degradeCounter[degradeLevel]++
+	}
+
+	volCounter := getOrCreateCounter(v.curVolStats)
+	volCounter.degradeCounter[maxMissingShard]++
+
+	span.Debugf("succeed to update degrade info: vid[%d], codemode[%s], bids[%d]",
+		vid, codeMode.String(), len(bidsMissed))
+}
+
+func (v *VolumeDegradeStatsMgr) getBatchIndex(vid uint32) uint32 {
+	return vid / uint32(v.cfg.Batch)
+}
+
+func (v *VolumeDegradeStatsMgr) initialize() error {
+	span, ctx := trace.StartSpanFromContext(context.Background(), "VolumeDegrade.Initialize")
+	defer span.Finish()
+
+	if !v.cfg.Enable {
+		span.Infof("volume degrade stats don't enabled, just return")
+		return nil
+	}
+
+	if err := v.getStatsFromKV(ctx); err != nil {
+		if !strings.Contains(err.Error(), errcode.ErrNotFound.Error()) {
+			span.Errorf("failed to get volume degrade batch size from kv, err[%+v]", err)
+			return err
+		}
+	}
+	span.Infof("succeed to get all exist stats from kv")
+
+	kvBatchSize := v.stats.batchSize
+	if kvBatchSize != v.cfg.Batch {
+		span.Infof("degrade batch changed from %d to %d, will delete all exist stats", kvBatchSize, v.cfg.Batch)
+		if err := v.deleteStatsFromKV(ctx); err != nil {
+			span.Errorf("failed to delete all volume degrade batch stats from kv, err[%+v]", err)
+			return err
+		}
+		span.Infof("succeed to delete all exist stats from clustermgr")
+		v.stats = newDegradeStatsCounter(v.cfg.Batch)
+	}
+
+	span.Infof("succeed to initialized: batchSize[%d], stats[%+v]", v.cfg.Batch, v.stats)
+	return nil
+}
+
+func (v *VolumeDegradeStatsMgr) setStatsToKV(ctx context.Context) error {
+	span := trace.SpanFromContextSafe(ctx)
+	record := v.stats.toRecord()
+	if err := v.clusterMgrCli.SetVolumeDegradeStats(ctx, record); err != nil {
+		span.Errorf("failed to set volume degrade stats[%+v] to kv, err[%+v]", record, err)
+		return err
+	}
+	span.Debugf("succeed to set volume degrade stats[%+v] to kv", record)
+	return nil
+}
+
+func (v *VolumeDegradeStatsMgr) getStatsFromKV(ctx context.Context) error {
+	span := trace.SpanFromContextSafe(ctx)
+	record, err := v.clusterMgrCli.GetVolumeDegradeStats(ctx)
+	if err != nil {
+		span.Errorf("failed to get volume degrade stats from kv, err[%+v]", err)
+		return err
+	}
+	span.Debugf("succeed to get volume degrade stats[%+v] from kv", record)
+	v.stats.fromRecord(record)
+	return nil
+}
+
+func (v *VolumeDegradeStatsMgr) deleteStatsFromKV(ctx context.Context) error {
+	span := trace.SpanFromContextSafe(ctx)
+	if err := v.clusterMgrCli.DeleteVolumeDegradeStats(ctx); err != nil {
+		span.Errorf("failed to delete volume degrade stats from kv, err[%+v]", err)
+		return err
+	}
+	span.Debugf("succeed to delete volume degrade stats from kv")
+	return nil
+}
+
+func aggregateDegradeStats(batches map[uint32]*degradeBatchCounter) map[codemode.CodeMode]map[int]float64 {
+	aggregated := make(map[codemode.CodeMode]map[int]float64)
+
+	for _, batch := range batches {
+		if batch == nil {
+			continue
+		}
+		for cm, counter := range batch.codeModeCounter {
+			if counter == nil {
+				continue
+			}
+			if _, ok := aggregated[cm]; !ok {
+				aggregated[cm] = make(map[int]float64)
+			}
+			for degradeLevel, count := range counter.degradeCounter {
+				aggregated[cm][degradeLevel] += float64(count)
+			}
+		}
+	}
+
+	return aggregated
+}
+
+func (v *VolumeDegradeStatsMgr) updateMetric() {
+	clusterID := v.clusterID.ToString()
+	blobAggregated := aggregateDegradeStats(v.stats.blobCounter)
+	volAggregated := aggregateDegradeStats(v.stats.volCounter)
+	reportDegradeStats(clusterID, _metricDegradeStatsItemBlob, blobAggregated)
+	reportDegradeStats(clusterID, _metricDegradeStatsItemVolume, volAggregated)
+}

--- a/blobstore/scheduler/volume_degrade_stats_test.go
+++ b/blobstore/scheduler/volume_degrade_stats_test.go
@@ -1,0 +1,349 @@
+// Copyright 2022 The CubeFS Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package scheduler
+
+import (
+	"context"
+	"encoding/json"
+	"math"
+	"testing"
+
+	"github.com/cubefs/cubefs/blobstore/common/codemode"
+	"github.com/cubefs/cubefs/blobstore/common/errors"
+	"github.com/cubefs/cubefs/blobstore/common/proto"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+)
+
+func newDefaultDegradeStats(t *testing.T) *VolumeDegradeStatsMgr {
+	ctrl := gomock.NewController(t)
+	mockCli := NewMockClusterMgrAPI(ctrl)
+	clusterID := proto.ClusterID(1)
+	cfg := &VolumeDegradeStatsCfg{
+		Enable: true,
+		Batch:  3,
+	}
+	return newVolumeDegradeStatsMgr(clusterID, mockCli, cfg)
+}
+
+func genVolumeDegradeStats(batchSize uint32, numBatches int) *proto.VolumeDegradeStats {
+	blobBatchs := make([]proto.VolumeDegradeBatch, numBatches)
+	for i := 0; i < numBatches; i++ {
+		batchIndex := uint32(i)
+		codeModeStats := []proto.VolumeDegradeCodeMode{
+			{
+				Mode: codemode.EC6P6,
+				DegradeStats: []proto.VolumeDegradeLevel{
+					{Level: 1, Count: 123},
+					{Level: 2, Count: 456},
+				},
+			},
+			{
+				Mode: codemode.EC3P3,
+				DegradeStats: []proto.VolumeDegradeLevel{
+					{Level: 3, Count: 789},
+				},
+			},
+		}
+		blobBatchs[i] = proto.VolumeDegradeBatch{
+			BatchIndex:    batchIndex,
+			CodeModeStats: codeModeStats,
+		}
+	}
+
+	volBatchs := make([]proto.VolumeDegradeBatch, numBatches)
+	for i := 0; i < numBatches; i++ {
+		batchIndex := uint32(i)
+		codeModeStats := []proto.VolumeDegradeCodeMode{
+			{
+				Mode: codemode.EC6P6,
+				DegradeStats: []proto.VolumeDegradeLevel{
+					{Level: 1, Count: 1},
+					{Level: 2, Count: 2},
+				},
+			},
+			{
+				Mode: codemode.EC3P3,
+				DegradeStats: []proto.VolumeDegradeLevel{
+					{Level: 3, Count: 3},
+				},
+			},
+		}
+		volBatchs[i] = proto.VolumeDegradeBatch{
+			BatchIndex:    batchIndex,
+			CodeModeStats: codeModeStats,
+		}
+	}
+
+	return &proto.VolumeDegradeStats{
+		BatchSize: batchSize,
+		BlobStats: blobBatchs,
+		VolStats:  volBatchs,
+	}
+}
+
+func genBidsMissed() map[proto.BlobID][]uint8 {
+	bids := make(map[proto.BlobID][]uint8)
+
+	{
+		bid := proto.BlobID(1)
+		shardCount := 1
+		shards := make([]uint8, shardCount)
+		for i := 0; i < shardCount; i++ {
+			shards[i] = uint8(i)
+		}
+		bids[bid] = shards
+	}
+
+	{
+		bid := proto.BlobID(2)
+		shardCount := 2
+		shards := make([]uint8, shardCount)
+		for i := 0; i < shardCount; i++ {
+			shards[i] = uint8(i)
+		}
+		bids[bid] = shards
+	}
+
+	{
+		bid := proto.BlobID(3)
+		shardCount := 3
+		shards := make([]uint8, shardCount)
+		for i := 0; i < shardCount; i++ {
+			shards[i] = uint8(i)
+		}
+		bids[bid] = shards
+	}
+
+	{
+		bid := proto.BlobID(4)
+		shardCount := 4
+		shards := make([]uint8, shardCount)
+		for i := 0; i < shardCount; i++ {
+			shards[i] = uint8(i)
+		}
+		bids[bid] = shards
+	}
+	return bids
+}
+
+func TestDegradeStatsInitialize(t *testing.T) {
+	// not enabled
+	{
+		mgr := newDefaultDegradeStats(t)
+		mgr.cfg.Enable = false
+		err := mgr.initialize()
+		require.NoError(t, err)
+	}
+	// get stats failed
+	{
+		mgr := newDefaultDegradeStats(t)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().GetVolumeDegradeStats(any).AnyTimes().Return(nil, errMock)
+
+		err := mgr.initialize()
+		require.NotNil(t, err)
+	}
+	// not found
+	{
+		mgr := newDefaultDegradeStats(t)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().GetVolumeDegradeStats(any).AnyTimes().Return(nil, errors.ErrNotFound)
+
+		err := mgr.initialize()
+		require.NoError(t, err)
+		require.Equal(t, uint32(3), mgr.stats.batchSize)
+		require.Equal(t, 0, len(mgr.stats.blobCounter))
+		require.Equal(t, 0, len(mgr.stats.volCounter))
+	}
+	// delete stats failed
+	{
+		mgr := newDefaultDegradeStats(t)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().GetVolumeDegradeStats(any).AnyTimes().Return(genVolumeDegradeStats(10, 3), nil)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().DeleteVolumeDegradeStats(any).AnyTimes().Return(errMock)
+
+		err := mgr.initialize()
+		require.NotNil(t, err)
+	}
+	// 3 batch stats
+	{
+		mgr := newDefaultDegradeStats(t)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().GetVolumeDegradeStats(any).AnyTimes().Return(genVolumeDegradeStats(3, 3), nil)
+
+		err := mgr.initialize()
+		require.NoError(t, err)
+		require.Equal(t, uint32(3), mgr.stats.batchSize)
+		require.Equal(t, 3, len(mgr.stats.blobCounter))
+		require.Equal(t, 3, len(mgr.stats.volCounter))
+	}
+	// new batch size
+	{
+		mgr := newDefaultDegradeStats(t)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().GetVolumeDegradeStats(any).AnyTimes().Return(genVolumeDegradeStats(10, 3), nil)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().DeleteVolumeDegradeStats(any).AnyTimes().Return(nil)
+
+		err := mgr.initialize()
+		require.NoError(t, err)
+		require.Equal(t, uint32(3), mgr.stats.batchSize)
+		require.Equal(t, 0, len(mgr.stats.blobCounter))
+		require.Equal(t, 0, len(mgr.stats.volCounter))
+	}
+}
+
+func TestDegradeStatsUpdate(t *testing.T) {
+	ctx := context.Background()
+	// new cluster
+	{
+		mgr := newDefaultDegradeStats(t)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().GetVolumeDegradeStats(any).AnyTimes().Return(nil, errors.ErrNotFound)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().SetVolumeDegradeStats(any, any).AnyTimes().Return(nil)
+
+		err := mgr.initialize()
+		require.NoError(t, err)
+		require.Equal(t, uint32(3), mgr.stats.batchSize)
+		require.Equal(t, 0, len(mgr.stats.blobCounter))
+		require.Equal(t, 0, len(mgr.stats.volCounter))
+
+		mgr.updateDegradeInfo(ctx, 0, codemode.EC3P3, genBidsMissed())
+		mgr.updateDegradeInfo(ctx, 1, codemode.EC3P3, genBidsMissed())
+		mgr.updateDegradeInfo(ctx, 2, codemode.EC3P3, genBidsMissed())
+		require.Equal(t, 0, len(mgr.stats.blobCounter))
+		require.Equal(t, 0, len(mgr.stats.volCounter))
+		require.Equal(t, uint32(0), mgr.curBlobStats.batchIndex)
+
+		mgr.updateDegradeInfo(ctx, 3, codemode.EC3P3, genBidsMissed())
+		require.Equal(t, 1, len(mgr.stats.blobCounter))
+		require.Equal(t, 1, len(mgr.stats.volCounter))
+		require.Equal(t, uint32(1), mgr.curBlobStats.batchIndex)
+	}
+	// set stats failed
+	{
+		mgr := newDefaultDegradeStats(t)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().GetVolumeDegradeStats(any).AnyTimes().Return(nil, errors.ErrNotFound)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().SetVolumeDegradeStats(any, any).AnyTimes().Return(errMock)
+
+		err := mgr.initialize()
+		require.NoError(t, err)
+		require.Equal(t, uint32(3), mgr.stats.batchSize)
+		require.Equal(t, 0, len(mgr.stats.blobCounter))
+		require.Equal(t, 0, len(mgr.stats.volCounter))
+
+		mgr.updateDegradeInfo(ctx, 0, codemode.EC3P3, genBidsMissed())
+		mgr.updateDegradeInfo(ctx, 1, codemode.EC3P3, genBidsMissed())
+		mgr.updateDegradeInfo(ctx, 2, codemode.EC3P3, genBidsMissed())
+		require.Equal(t, 0, len(mgr.stats.blobCounter))
+		require.Equal(t, 0, len(mgr.stats.volCounter))
+		require.Equal(t, uint32(0), mgr.curBlobStats.batchIndex)
+
+		mgr.updateDegradeInfo(ctx, 3, codemode.EC3P3, genBidsMissed())
+		require.Equal(t, 1, len(mgr.stats.blobCounter))
+		require.Equal(t, 1, len(mgr.stats.volCounter))
+		require.Equal(t, uint32(1), mgr.curBlobStats.batchIndex)
+	}
+	// service restart
+	{
+		mgr := newDefaultDegradeStats(t)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().GetVolumeDegradeStats(any).AnyTimes().Return(genVolumeDegradeStats(3, 3), nil)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().SetVolumeDegradeStats(any, any).AnyTimes().Return(nil)
+
+		err := mgr.initialize()
+		require.Nil(t, err)
+		require.Equal(t, uint32(3), mgr.stats.batchSize)
+		require.Equal(t, 3, len(mgr.stats.blobCounter))
+		require.Equal(t, 3, len(mgr.stats.volCounter))
+
+		mgr.updateDegradeInfo(ctx, 3000, codemode.EC3P3, genBidsMissed())
+		mgr.updateDegradeInfo(ctx, 3001, codemode.EC3P3, genBidsMissed())
+		mgr.updateDegradeInfo(ctx, 3002, codemode.EC3P3, genBidsMissed())
+		require.Equal(t, 3, len(mgr.stats.blobCounter))
+		require.Equal(t, 3, len(mgr.stats.volCounter))
+
+		mgr.updateDegradeInfo(ctx, 30003, codemode.EC3P3, genBidsMissed())
+		require.Equal(t, 4, len(mgr.stats.blobCounter))
+		require.Equal(t, 4, len(mgr.stats.volCounter))
+	}
+	// service restart with new batchSize
+	{
+		mgr := newDefaultDegradeStats(t)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().GetVolumeDegradeStats(any).AnyTimes().Return(genVolumeDegradeStats(10, 3), nil)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().DeleteVolumeDegradeStats(any).AnyTimes().Return(nil)
+		mgr.clusterMgrCli.(*MockClusterMgrAPI).EXPECT().SetVolumeDegradeStats(any, any).AnyTimes().Return(nil)
+
+		err := mgr.initialize()
+		require.Nil(t, err)
+		require.Equal(t, uint32(3), mgr.stats.batchSize)
+		require.Equal(t, 0, len(mgr.stats.blobCounter))
+		require.Equal(t, 0, len(mgr.stats.volCounter))
+
+		mgr.updateDegradeInfo(ctx, 3000, codemode.EC3P3, genBidsMissed())
+		mgr.updateDegradeInfo(ctx, 3001, codemode.EC3P3, genBidsMissed())
+		mgr.updateDegradeInfo(ctx, 3002, codemode.EC3P3, genBidsMissed())
+		require.Equal(t, 0, len(mgr.stats.blobCounter))
+		require.Equal(t, 0, len(mgr.stats.volCounter))
+
+		mgr.updateDegradeInfo(ctx, 30003, codemode.EC3P3, genBidsMissed())
+		require.Equal(t, 1, len(mgr.stats.blobCounter))
+		require.Equal(t, 1, len(mgr.stats.volCounter))
+	}
+}
+
+func TestDegradeStatsKVSize(t *testing.T) {
+	volumeTotalCount := uint32(2000000)
+	batchSize := uint32(10000)
+	numBatches := volumeTotalCount / batchSize
+
+	buildBatches := func() []proto.VolumeDegradeBatch {
+		batches := make([]proto.VolumeDegradeBatch, 0, numBatches)
+		for i := uint32(0); i < numBatches; i++ {
+			var cm codemode.CodeMode
+			if i < numBatches/2 {
+				cm = codemode.EC15P12
+			} else {
+				cm = codemode.EC12P9
+			}
+			totalShard := cm.Tactic().N + cm.Tactic().M
+
+			degradeStats := make([]proto.VolumeDegradeLevel, totalShard)
+			for j := 1; j <= totalShard; j++ {
+				degradeStats[j-1] = proto.VolumeDegradeLevel{
+					Level: j,
+					Count: math.MaxUint32,
+				}
+			}
+
+			batches = append(batches, proto.VolumeDegradeBatch{
+				BatchIndex: i,
+				CodeModeStats: []proto.VolumeDegradeCodeMode{
+					{
+						Mode:         cm,
+						DegradeStats: degradeStats,
+					},
+				},
+			})
+		}
+		return batches
+	}
+
+	blobBatchs := buildBatches()
+	volBatchs := buildBatches()
+
+	stats := &proto.VolumeDegradeStats{
+		BatchSize: batchSize,
+		BlobStats: blobBatchs,
+		VolStats:  volBatchs,
+	}
+
+	statsBytes, err := json.Marshal(stats)
+	require.NoError(t, err)
+	require.LessOrEqual(t, len(statsBytes), 1024*1024)
+}

--- a/blobstore/scheduler/volume_inspector.go
+++ b/blobstore/scheduler/volume_inspector.go
@@ -147,6 +147,8 @@ func (d *badShardDeduplicator) key(vid proto.Vid, bid proto.BlobID, badIdxs []ui
 
 // VolumeInspectMgrCfg inspect task manager config
 type VolumeInspectMgrCfg struct {
+	ClusterID proto.ClusterID `json:"-"`
+
 	InspectIntervalS int `json:"inspect_interval_s"`
 	InspectBatch     int `json:"inspect_batch"`
 
@@ -156,6 +158,9 @@ type VolumeInspectMgrCfg struct {
 
 	// timeout of inspect
 	TimeoutMs int `json:"timeout_ms"`
+
+	// Degrade stats
+	DegradeStats VolumeDegradeStatsCfg `json:"degrade_stats"`
 }
 
 // VolumeInspectMgr inspect task manager
@@ -183,6 +188,8 @@ type VolumeInspectMgr struct {
 	completeTaskCounter counter.Counter
 	timeoutCounter      counter.Counter
 
+	degradeStats *VolumeDegradeStatsMgr
+
 	cfg *VolumeInspectMgrCfg
 }
 
@@ -201,6 +208,7 @@ func NewVolumeInspectMgr(
 		clusterMgrCli:     clusterMgrCli,
 		repairShardSender: repairShardSender,
 		sendDeduplicator:  newBadShardDeduplicator(defaultDuplicateCnt),
+		degradeStats:      newVolumeDegradeStatsMgr(cfg.ClusterID, clusterMgrCli, &cfg.DegradeStats),
 		cfg:               cfg,
 	}
 }
@@ -422,6 +430,13 @@ func (mgr *VolumeInspectMgr) finish(ctx context.Context) {
 		}
 	}
 
+	// sort the missingShards to ensure that the degradeStats batch processing works correctly.
+	sort.Slice(missedShards, func(i, j int) bool {
+		vidI := missedShards[i][0].Vuid.Vid()
+		vidJ := missedShards[j][0].Vuid.Vid()
+		return vidI < vidJ
+	})
+
 	// clear & stats tasks
 	for taskID, task := range mgr.tasks {
 		span.Debugf("check task and clear: task_id[%s]", taskID)
@@ -463,6 +478,8 @@ func (mgr *VolumeInspectMgr) finish(ctx context.Context) {
 				return mgr.trySendShardRepairMsg(ctx, vid, bid, bads)
 			})
 		}
+
+		mgr.degradeStats.updateDegradeInfo(ctx, uint32(vid), volInfo.CodeMode, bidsBads)
 	}
 
 	err := retry.Timed(3, 200).On(func() error {

--- a/docs-zh/source/ops/configs/blobstore/scheduler.md
+++ b/docs-zh/source/ops/configs/blobstore/scheduler.md
@@ -185,15 +185,24 @@ v3.3.0版本开始支持并发修复磁盘。
 * list_vol_step，请求clustermgr列举卷大小，可控制请求clustermgr的qps，默认100
 * list_vol_interval_ms，请求clustermgr列举卷的时间间隔，默认10ms
 * timeout_ms，检查一批巡检任务是否完成的时间间隔，默认10000ms
+* degrade_stats，降级状态统计
+  * enable，是否启用 volume 降级统计功能，默认false
+  * batch，批量统计大小，默认10000，最好是inspect_batch的倍数
+
 ```json
 {
-    "inspect_interval_s": 100,     
-    "inspect_batch": 10,    
-    "list_vol_step": 20,    
-    "list_vol_interval_ms": 10,    
-    "timeout_ms": 10000   
+    "inspect_interval_s": 100,
+    "inspect_batch": 10,
+    "list_vol_step": 20,
+    "list_vol_interval_ms": 10,
+    "timeout_ms": 10000,
+    "degrade_stats": {
+        "enable": true,
+        "batch": 1000
+    }
 }
 ```
+
 ### shard_repair示例
 
 * task_pool_size，修补任务的并发度，默认10

--- a/docs/source/ops/configs/blobstore/scheduler.md
+++ b/docs/source/ops/configs/blobstore/scheduler.md
@@ -187,15 +187,24 @@ Starting from version v3.3.0, concurrent disk repair is supported.
 * list_vol_step, the size of requesting clustermgr to list volumes, which can control the QPS of requesting clustermgr, default is 100
 * list_vol_interval_ms, time interval for requesting clustermgr to list volumes, default is 10ms
 * timeout_ms, time interval for checking whether a batch of inspection tasks is completed, default is 10000ms
+* degrade_stats, volume degrade stats
+  * enable: enable volume degrade stats or not, default is false
+  * batch: Batch size for statistics, defaults to 10000, ideally a multiple of inspect_batch
+
 ```json
 {
-    "inspect_interval_s": 100,     
-    "inspect_batch": 10,    
-    "list_vol_step": 20,    
-    "list_vol_interval_ms": 10,    
-    "timeout_ms": 10000   
+    "inspect_interval_s": 100,
+    "inspect_batch": 10,
+    "list_vol_step": 20,
+    "list_vol_interval_ms": 10,
+    "timeout_ms": 10000,
+    "degrade_stats": {
+        "enable": true,
+        "batch": 1000
+    }
 }
 ```
+
 ### shard_repair
 
 * task_pool_size, concurrency of repair tasks, default is 10

